core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -213,37 +210,39 @@ public final CompletableFuture<List<Endpoint>> whenReady() {
      * Adds the specified {@link Endpoint} to current {@link Endpoint} list.
      */
     protected final void addEndpoint(Endpoint e) {
-        final List<Endpoint> newEndpoints;
-        endpointsLock.lock();
+        lock();
         try {
             final List<Endpoint> newEndpointsUnsorted = Lists.newArrayList(endpoints);
             newEndpointsUnsorted.add(e);
-            endpoints = newEndpoints = ImmutableList.sortedCopyOf(newEndpointsUnsorted);
+            final List<Endpoint> newEndpoints = ImmutableList.sortedCopyOf(newEndpointsUnsorted);
+            unsafeUpdateEndpoints(newEndpoints);
         } finally {
-            endpointsLock.unlock();
+            unlock();
         }
+    }
 
-        notifyListeners(newEndpoints);
-        completeInitialEndpointsFuture(newEndpoints);
+    @Override
+    protected final List<Endpoint> latestValue() {
```

Comment:
Question) is this override still necessary?

From what I understood, if `setEndpoints` is called before or after `endpointGroup#addListener` then we shouldn't have a race condition anymore.

https://github.com/line/armeria/blob/1c93feef523b619d1d73109772fe1530fb0f110e/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L76-L76

Reply:
Right. Technically, we don't have to override it since the race condition has been gone. However, I didn't revert it since the new version is slightly optimized.